### DISCUSSION TITLE
Content/Parameters is missing

### DISCUSSION THREAD

**Discussion by JoaoBrito0** [Cornutum/tcases#273] [General]
(get).parameters.[name].content is missing
(get).responses.200.schema is unexpected
(post).consumes is unexpected

I'm trying to run tcases on my apis, but all of my endpoints return errors like so. Can you guys help trying to understand what could I be doing wrong?

**Comment by kerrykimbrough** [comment]
Happy to help but I need more information. 

When do these error messages appear? When you try to generate the test code? When the test code executes? When validating request responses?

Also, please attach a copy of your OpenAPI definition.

**Reply by JoaoBrito0** [reply]
these error messages appear when I run tcases-api <my swagger json document>

**Reply by kerrykimbrough** [reply]
If you question has been resolved, please mark the reply below as the answer.

**Comment by kerrykimbrough** [comment]
If I had the OpenAPI definition, I could tell you more. But I'm fairly sure these messages are reporting errors in your OpenAPI definition.  You can consult with the [OpenAPI specification](https://spec.openapis.org/oas/v3.0.2) to learn what's going wrong.

**Reply by trogdor259** [reply]
I am waiting on permission from higher ups to post the doc here (it's internal for work). In the meantime, does tcases support using OpenAPI 3.0.1? When I use swagger.io to convert a Swagger 2.0 file to OpenAPI 3.0.1 format the tcases-api does not work at all. When I use a different tool to convert it to OpenAPI 3.0.0 I get the results I have been seeing. Could this be part of the reason why the parameters are not being generated?

**Reply by kerrykimbrough** [reply]
tcases-api was designed based OpenAPI 3.0.2. I have not tested it with 3.0.0 or 3.0.1, so it is possible there could be problems. I would need to look into what the differences in these versions are.

At any rate, I'm sure any changes needed to make your definition compatible with 3.0.2 will be small.

**Reply by kerrykimbrough** [reply]
Looking at the [OAS release notes](https://github.com/OAI/OpenAPI-Specification/releases), it appears there are no functional differences between any of the 3.0.x versions of the spec. Of course, there still could be problems introduced by the 2.x to 3.0.x migration process.

**Comment by trogdor259** [comment]
I am having a similar situation. I have been looking at the logs when generating the cases, but I am not seeing any errors there (unless I am missing something). I can pm you my definition if you'd like.

**Comment by kerrykimbrough** [comment]
Yes, please send your OpenAPI definition for investigation.

**Reply by trogdor259** [reply]
Sorry for the delay. I had to sanitize the data to share here. I have removed many of the endpoints, but there should be sufficient here for you.

`{
  "openapi": "3.0.1",
  "info": {
    "title": "RulePublic API",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "/"
    }
  ],
  "paths": {
    "/cloudrules/v2/systems/{systemID}/routines": {
      "get": {
        "tags": [
          "RulePublic"
        ],
        "summary": "Gets all the routines for a system",
        "description": "Returns all the routines that are tied to the specified system",
        "operationId": "RulePublic_GetAllRoutinesBySystemID",
        "parameters": [
          {
            "name": "systemID",
            "in": "path",
            "description": "ID of the system the routine is on",
            "required": true,
            "schema": {
              "type": "string",
              "format": "int64"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "A successful response.",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/dataRoutineList"
                }
              }
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/runtimeError"
                }
              }
            }
          }
        }
      },
      "post": {
        "tags": [
          "RulePublic"
        ],
        "summary": "Creates a routine for the specified system",
        "description": "Returns all the routines including the newly created one on success",
        "operationId": "RulePublic_CreateRoutine",
        "parameters": [
          {
            "name": "systemID",
            "in": "path",
            "description": "The id of the system the routine is associated with",
            "required": true,
            "schema": {
              "type": "string",
              "format": "int64"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/dataRoutineEditableFields"
              }
            }
          },
          "required": true
        },
        "responses": {
          "200": {
            "description": "A successful response.",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/dataRoutineList"
                }
              }
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/runtimeError"
                }
              }
            }
          }
        },
        "x-codegen-request-body-name": "body"
      }
    }
  },
  "components": {
    "schemas": {
      "dataActionList": {
        "type": "object",
        "properties": {
          "actions": {
            "type": "array",
            "description": "A list of actions",
            "items": {
              "$ref": "#/components/schemas/dataRoutineAction"
            }
          }
        }
      },
      "dataCarGuardLocation": {
        "type": "object",
        "properties": {
          "id": {
            "type": "string"
          },
          "n": {
            "type": "string"
          }
        }
      },
      "dataConditionList": {
        "type": "object",
        "properties": {
          "conditions": {
            "type": "array",
            "description": "A list of conditions",
            "items": {
              "$ref": "#/components/schemas/dataRoutineCondition"
            }
          }
        }
      },
      "dataDevice": {
        "type": "object",
        "properties": {
          "id": {
            "type": "integer",
            "description": "The id of the device",
            "format": "int64"
          },
          "s": {
            "type": "integer",
            "description": "Status",
            "format": "int32"
          },
          "hsp": {
            "type": "number",
            "description": "Set point",
            "format": "float"
          },
          "csp": {
            "type": "number",
            "description": "Set point",
            "format": "float"
          },
          "n": {
            "type": "string",
            "description": "The name of the device"
          },
          "cpri": {
            "type": "boolean",
            "description": "Setting for the device"
          },
          "fm": {
            "type": "integer",
            "description": "Mode",
            "format": "int32"
          },
          "fd": {
            "type": "integer",
            "description": "The duration",
            "format": "int64"
          },
          "val": {
            "type": "integer",
            "description": "Value for the device",
            "format": "int32"
          },
          "t": {
            "type": "string",
            "description": "Device type"
          },
          "ol": {
            "type": "boolean",
            "description": "The Online status of the device"
          }
        }
      },
      "dataExpectedDeviceChanges": {
        "type": "object",
        "properties": {
          "devices": {
            "type": "array",
            "description": "The set of devices that are expected to be modified",
            "items": {
              "$ref": "#/components/schemas/dataDevice"
            }
          }
        }
      },
      "dataRestrictions": {
        "type": "object",
        "properties": {
          "id": {
            "type": "integer",
            "description": "The id",
            "format": "int32"
          },
          "s": {
            "type": "array",
            "description": "The status",
            "items": {
              "type": "integer",
              "format": "int32"
            }
          }
        }
      },
      "dataRoutine": {
        "type": "object",
        "properties": {
          "id": {
            "$ref": "#/components/schemas/objectidObjectIdMessage"
          },
          "name": {
            "type": "string",
            "description": "The name"
          },
          "systemId": {
            "type": "string",
            "description": "Id",
            "format": "int64"
          },
          "systemType": {
            "type": "integer",
            "description": "The type of system",
            "format": "int32"
          },
          "category": {
            "type": "string",
            "description": "The category"
          },
          "createdTs": {
            "$ref": "#/components/schemas/timeTimeMessage"
          },
          "actions": {
            "type": "array",
            "description": "Actions",
            "items": {
              "$ref": "#/components/schemas/dataRoutineAction"
            }
          },
          "icon": {
            "type": "string",
            "description": "The icon"
          },
          "description": {
            "type": "string",
            "description": "The description"
          },
          "triggers": {
            "type": "array",
            "description": "Triggers",
            "items": {
              "$ref": "#/components/schemas/dataRoutineTrigger"
            }
          },
          "enabled": {
            "type": "boolean",
            "description": "Enabled flag"
          },
          "isShortcut": {
            "type": "boolean",
            "description": "Flag to set"
          },
          "conditions": {
            "type": "array",
            "description": "Confitions",
            "items": {
              "$ref": "#/components/schemas/dataRoutineCondition"
            }
          },
          "accountSystemId": {
            "type": "string",
            "description": "Id",
            "format": "int64"
          }
        }
      },
      "dataRoutineAction": {
        "type": "object",
        "properties": {
          "id": {
            "type": "integer",
            "description": "Id",
            "format": "int32"
          },
          "name": {
            "type": "string",
            "description": "Name"
          },
          "icon": {
            "type": "string",
            "description": "Icon"
          },
          "category": {
            "type": "string",
            "description": "Category"
          },
          "devices": {
            "type": "array",
            "description": "Devices",
            "items": {
              "$ref": "#/components/schemas/dataDevice"
            }
          },
          "toggle_time": {
            "$ref": "#/components/schemas/dataRoutineTimeParam"
          }
        }
      },
      "dataRoutineActionEditableFields": {
        "type": "object",
        "properties": {
          "devices": {
            "type": "array",
            "description": "Devices",
            "items": {
              "$ref": "#/components/schemas/dataDevice"
            }
          }
        }
      },
      "dataRoutineCondition": {
        "type": "object",
        "properties": {
          "id": {
            "type": "integer",
            "description": "Id",
            "format": "int32"
          },
          "name": {
            "type": "string",
            "description": "Name"
          },
          "params": {
            "$ref": "#/components/schemas/dataRoutineConditionParams"
          },
          "category": {
            "type": "string",
            "description": "Category"
          },
          "devices": {
            "type": "array",
            "description": "Devices",
            "items": {
              "$ref": "#/components/schemas/dataDevice"
            }
          }
        }
      },
      "dataRoutineConditionEditableFields": {
        "type": "object",
        "properties": {
          "params": {
            "$ref": "#/components/schemas/dataRoutineConditionParams"
          }
        }
      },
      "dataRoutineConditionParams": {
        "type": "object",
        "properties": {
          "startTime": {
            "$ref": "#/components/schemas/dataRoutineTimeParam"
          },
          "endTime": {
            "$ref": "#/components/schemas/dataRoutineTimeParam"
          },
          "standaloneDevicesData": {
            "$ref": "#/components/schemas/dataStandaloneDevicesData"
          }
        }
      },
      "dataRoutineEditableFields": {
        "title": "Nullable fields for the purpose of handling updates",
        "type": "object",
        "properties": {
          "name": {
            "type": "string",
            "description": "Name"
          },
          "category": {
            "type": "string",
            "description": "Category"
          },
          "actions": {
            "type": "array",
            "description": "Actions",
            "items": {
              "$ref": "#/components/schemas/dataRoutineAction"
            }
          },
          "triggers": {
            "type": "array",
            "description": "Triggers",
            "items": {
              "$ref": "#/components/schemas/dataRoutineTrigger"
            }
          },
          "enabled": {
            "type": "boolean",
            "description": "Enable/Disable"
          },
          "isShortcut": {
            "type": "boolean",
            "description": "is shortcut"
          },
          "conditions": {
            "type": "array",
            "description": "Conditions",
            "items": {
              "$ref": "#/components/schemas/dataRoutineCondition"
            }
          },
          "systemType": {
            "type": "integer",
            "description": "System Type",
            "format": "int32"
          }
        }
      },
      "dataRoutineList": {
        "type": "object",
        "properties": {
          "routines": {
            "type": "array",
            "description": "Items",
            "items": {
              "$ref": "#/components/schemas/dataRoutine"
            }
          }
        }
      },
      "dataRoutineTimeParam": {
        "type": "object",
        "properties": {
          "hour": {
            "type": "integer",
            "description": "The hour (0-23)",
            "format": "int64"
          },
          "minute": {
            "type": "integer",
            "description": "The minute (0-59)",
            "format": "int64"
          },
          "special": {
            "type": "string",
            "description": "Special time (sunrise or sunset)"
          },
          "duration": {
            "type": "integer",
            "description": "A time (in seconds)",
            "format": "int64"
          }
        }
      },
      "dataRoutineTrigger": {
        "type": "object",
        "properties": {
          "id": {
            "type": "integer",
            "description": "Id",
            "format": "int32"
          },
          "name": {
            "type": "string",
            "description": "Name"
          },
          "params": {
            "$ref": "#/components/schemas/dataRoutineTriggerParams"
          },
          "category": {
            "type": "string",
            "description": "Category"
          },
          "restrictions": {
            "type": "array",
            "description": "Restrictions",
            "items": {
              "$ref": "#/components/schemas/dataRestrictions"
            }
          },
          "devices": {
            "type": "array",
            "description": "Devices",
            "items": {
              "$ref": "#/components/schemas/dataDevice"
            }
          },
          "standaloneDevicesData": {
            "$ref": "#/components/schemas/dataStandaloneDevicesData"
          }
        }
      },
      "dataRoutineTriggerEditableFields": {
        "type": "object",
        "properties": {
          "params": {
            "$ref": "#/components/schemas/dataRoutineTriggerParams"
          },
          "devices": {
            "type": "array",
            "description": "Devices",
            "items": {
              "$ref": "#/components/schemas/dataDevice"
            }
          },
          "standaloneDevicesData": {
            "$ref": "#/components/schemas/dataStandaloneDevicesData"
          }
        }
      },
      "dataRoutineTriggerParams": {
        "type": "object",
        "properties": {
          "time": {
            "$ref": "#/components/schemas/dataRoutineTimeParam"
          },
          "days": {
            "type": "array",
            "description": "Days",
            "items": {
              "type": "integer",
              "format": "int64"
            }
          }
        }
      },
      "dataStandaloneDevicesData": {
        "title": "contains data for standalone triggers, add new fields as needed",
        "type": "object",
        "properties": {
          "vehicles": {
            "type": "array",
            "description": "Vehicles",
            "items": {
              "$ref": "#/components/schemas/dataVehicle"
            }
          },
          "locations": {
            "type": "array",
            "description": "Locations",
            "items": {
              "$ref": "#/components/schemas/dataCarGuardLocation"
            }
          },
          "shsID": {
            "type": "string",
            "description": "Id",
            "format": "int64"
          }
        }
      },
      "dataTriggerList": {
        "type": "object",
        "properties": {
          "triggers": {
            "type": "array",
            "description": "Triggers",
            "items": {
              "$ref": "#/components/schemas/dataRoutineTrigger"
            }
          }
        }
      },
      "dataVehicle": {
        "type": "object",
        "properties": {
          "id": {
            "type": "string"
          },
          "n": {
            "type": "string"
          }
        }
      },
      "objectidObjectIdMessage": {
        "title": "MessageId",
        "type": "object",
        "properties": {
          "objectId": {
            "type": "string"
          }
        }
      },
      "protobufAny": {
        "type": "object",
        "properties": {
          "typeUrl": {
            "type": "string",
            "description": "typeURL"
          },
          "value": {
            "pattern": "^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$",
            "type": "string",
            "description": "Value",
            "format": "byte"
          }
        },
        "description": "Any"
      },
      "rulepublicRoutineEditableFieldsList": {
        "type": "object",
        "properties": {
          "routines": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/dataRoutineEditableFields"
            }
          }
        }
      },
      "runtimeError": {
        "type": "object",
        "properties": {
          "error": {
            "type": "string"
          },
          "code": {
            "type": "integer",
            "format": "int32"
          },
          "message": {
            "type": "string"
          },
          "details": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/protobufAny"
            }
          }
        }
      },
      "timeTimeMessage": {
        "type": "object",
        "properties": {
          "timeString": {
            "type": "string"
          }
        }
      }
    }
  }
}`

**Reply by kerrykimbrough** [reply]
When I run the latest version of `tcases-api` and `tcases-api-test` (4.0.1) on this definition, a result is produced with no errors.

Are you running a different version? Can you reproduce your problem with this definition? If so, can you send me any log files or error messages that are produced?

**Reply by trogdor259** [reply]
When I run it I do not get any errors. I am looking to get the pattern created for the payloads. Like the values that should be run in the tests. I see them when I run the examples in the docs provided, but not in any of the cases created with the json I am passing in. The example case looks like this:

`"paths": {
         "/allOf": {
             "post": {
                 "parameters": [
                     {
                         "name": "param0",
                         "in": "header",

                         "schema": {
                             "allOf": [
                                 {
                                     "type": "string",
                                     "pattern": ".*",
                                     "anyOf": [
                                         {
                                             "enum": [ "A", "B", "C" ]
                                         },
                                         {
                                             "minLength": 2
                                         },
                                         {
                                             "type": "boolean"
                                         }
                                     ],
                                     "oneOf": [
                                         {
                                             "type": "object"
                                         },
                                         {
                                             "pattern": "XXX"
                                         },
                                         {
                                             "maxLength": 1
                                         }
                                     ]
                                 }
                             ]
                         }
                     }
                 ],`
What I am producing does not have the pattern anywhere.

**Reply by trogdor259** [reply]
I am using Tcases 4.0.1 (2023-01-16)

**Reply by kerrykimbrough** [reply]
I'm a little confused now. You seem to be describing a problem that is different from the original subject of this discussion. And I'm not sure I understand what you think is going wrong. In this definition, the only instance of a "pattern" appears in the definition of `protobufAny`, which is used only in the definition of request responses. For example, if you look in the `RulePublicApiTest-Responses.json` file that was generated, you will see this pattern in the definition for  "An unexpected error response.".

I recommend you create a new issue or discussion with a complete explanation of what you attempted and what result you produced that you think is unexpected or incorrect. Also include any info needed to reproduce the problem.

**Comment by trogdor259** [comment]
Will do, thank you

On Tue, Mar 28, 2023 at 12:03 PM Kerry Kimbrough ***@***.***>
wrote:

> I'm a little confused now. You seem to be describing a problem that is
> different from the original subject of this discussion. And I'm not sure I
> understand what you think is going wrong. In this definition, the only
> instance of a "pattern" appears in the definition of protobufAny, which
> is used only in the definition of request responses. For example, if you
> look in the RulePublicApiTest-Responses.json file that was generated, you
> will see this pattern in the definition for "An unexpected error response.".
>
> I recommend you create a new issue or discussion with a complete
> explanation of what you attempted and what result you produced that you
> think is unexpected or incorrect. Also include any info needed to reproduce
> the problem.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Cornutum/tcases/discussions/273#discussioncomment-5457294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6ASSJSQMF6MYBMS4LS2H3W6MRYNANCNFSM6AAAAAAUIZ7JAQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
--
